File: project/suchan-onboarding/application/src/test/java/com/survey/application/ui/SurveyControllerTest.java
package com.survey.application.ui;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.survey.application.dto.dto.UpdateInputFormDto;
import com.survey.application.dto.dto.UpdateSurveyOptionDto;
import com.survey.application.dto.request.CreateSurveyRequest;
import com.survey.application.dto.request.UpdateSurveyRequest;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.http.MediaType;
import org.springframework.test.web.servlet.MockMvc;

import java.util.Collections;
import java.util.List;

import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.put;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

@SpringBootTest
@AutoConfigureMockMvc
class SurveyControllerTest {

    @Autowired
    private MockMvc mockMvc;

    @Autowired
    private ObjectMapper objectMapper;

    @Test
    @DisplayName("유효하지 않은 설문 조사 생성 요청이 전달되면 400 Bad Request와 validation error 메시지를 반환한다")
    void create_survey_validation_exception() throws Exception {
        // given
        CreateSurveyRequest invalidRequest = new CreateSurveyRequest(
                null,
                "설명",
                Collections.emptyList()
        );

        // when // then
        mockMvc.perform(post("/api/survey")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(objectMapper.writeValueAsString(invalidRequest)))
                .andExpect(status().isBadRequest())
                .andExpect(jsonPath("$.errorMessage").value("validation error"))
                .andExpect(jsonPath("$.errors.title").exists())
                .andExpect(jsonPath("$.errors.surveyOptionDtos").exists())
                .andExpect(jsonPath("$.timestamp").exists());
    }

    @Test
    @DisplayName("유효하지 않은 설문조사 수정 요청이 전달되면 400 Bad Request와 validation error 메시지를 반환한다")
    void update_survey_validation_exception() throws Exception {
        // given
        UpdateSurveyRequest invalidRequest = new UpdateSurveyRequest(
                1L,
                null,
                null,
                List.of(new UpdateSurveyOptionDto(
                        null,
                        "제목",
                        "설명",
                        true,
                        new UpdateInputFormDto(null, null, null)
                ))
        );

        // when // then
        mockMvc.perform(put("/api/survey")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(objectMapper.writeValueAsString(invalidRequest)))
                .andExpect(status().isBadRequest())
                .andExpect(jsonPath("$.errorMessage").value("validation error"))
                .andExpect(jsonPath("$.errors.title").value("must not be null"))
                .andExpect(jsonPath("$.errors.description").value("must not be null"))
                .andExpect(jsonPath("$.timestamp").exists());
    }

    @Test
    @DisplayName("SurveyOptionDto의 유효성 검사 실패 시 적절한 오류 메시지를 반환한다")
    void validation_exception() throws Exception {
        // given
        UpdateSurveyRequest invalidOptionRequest = new UpdateSurveyRequest(
                1L,
                "유효한 제목",
                "유효한 설명",
                List.of(new UpdateSurveyOptionDto(
                        null,
                        null,
                        null,
                        true,
                        new UpdateInputFormDto(
                                null,
                                null,
                                null
                        )
                ))
        );

        // when // then
        mockMvc.perform(put("/api/survey")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(objectMapper.writeValueAsString(invalidOptionRequest)))
                .andExpect(status().isBadRequest())
                .andExpect(jsonPath("$.errorMessage").value("validation error"))
                .andExpect(jsonPath("$.errors['surveyOptionDtos[0].title']").exists())
                .andExpect(jsonPath("$.errors['surveyOptionDtos[0].description']").exists())
                .andExpect(jsonPath("$.errors['surveyOptionDtos[0].inputFormDto.question']").exists());
    }

}

File: project/wonjun-onboarding/src/test/kotlin/com/onboarding/form/controller/SurveyControllerTest.kt
package com.onboarding.form.controller

import com.fasterxml.jackson.databind.ObjectMapper
import com.onboarding.form.domain.QuestionType
import com.onboarding.form.request.CreateSelectQuestionDto
import com.onboarding.form.request.CreateStandardQuestionDto
import com.onboarding.form.request.CreateSurveyDto
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.http.MediaType
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders
import org.springframework.test.web.servlet.result.MockMvcResultMatchers
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.status

@SpringBootTest
@AutoConfigureMockMvc
class SurveyControllerTest{
    @Autowired
    lateinit var mockMvc: MockMvc
    @Autowired
    lateinit var objectMapper: ObjectMapper

    @Test
    @DisplayName("Controller 요청시 성공 응답이 내려오는지 테스트")
    fun createSurvey(){
        val createSurveyDto = CreateSurveyDto(
            title = "testTitle",
            description = "testDescription",
            questions =  listOf(
                CreateSelectQuestionDto(
                    title = "testTitle",
                    description = "testDescription",
                    type = QuestionType.MULTI_SELECT,
                    isRequired = true,
                    answerList =  listOf("testAnswer1", "testAnswer2")
                ),
                CreateSelectQuestionDto(
                    title = "testTitle",
                    description = "testDescription",
                    type = QuestionType.SINGLE_SELECT,
                    isRequired = true,
                    answerList =  listOf("testAnswer1", "testAnswer2")
                ),
                CreateStandardQuestionDto(
                    title = "testTitle",
                    description = "testDescription",
                    type = QuestionType.SHORT,
                    isRequired = true,
                ),
                CreateStandardQuestionDto(
                    title = "testTitle",
                    description = "testDescription",
                    type = QuestionType.LONG,
                    isRequired = true,
                )
            )
        )

        mockMvc.perform(
            MockMvcRequestBuilders.post("/surveys")
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(createSurveyDto))
        ).andExpect(status().isOk)
            .andExpect(MockMvcResultMatchers.content().contentType(MediaType.APPLICATION_JSON))
            .andExpect(
                jsonPath("$.title").value(createSurveyDto.title)
            )
    }

    @Test
    @DisplayName("Survey에 Question을 10개이상 입력한 경우 에러 응답이 정상적으로 내려오는지 테스트")
    fun createSurveyFail(){
        val createSurveyDto = CreateSurveyDto(
            title = "testTitle",
            description = "testDescription",
            questions = (0 until 11).map { CreateStandardQuestionDto(
                title = "testTitle",
                description = "testDescription",
                type = QuestionType.LONG,
                isRequired = true,
            ) }.toList()
        )

        mockMvc.perform(
            MockMvcRequestBuilders.post("/surveys")
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(createSurveyDto))
        ).andExpect(status().isBadRequest)
            .andExpect(MockMvcResultMatchers.content().contentType(MediaType.APPLICATION_JSON))
            .andExpect(
                jsonPath("$.message").value("The number of questions cannot exceed 10")
            )
    }
}